cpp-client/deephaven/dhclient/src/arrowutil/arrow_array_converter.cc

```diff
@@ -0,0 +1,297 @@
+  OkOrThrow(DEEPHAVEN_LOCATION_EXPR(v.chunked_array_->type()->Accept(&v)));
+  return std::move(v.result_);
+}
+
+
+//-------------------------------------------------------------------------------------------
+//-------------------------------------------------------------------------------------------
+//-------------------------------------------------------------------------------------------
+//-------------------------------------------------------------------------------------------
+
+namespace {
+struct ColumnSourceToArrayVisitor final : ColumnSourceVisitor {
+  explicit ColumnSourceToArrayVisitor(size_t num_rows) : num_rows_(num_rows),
+      row_sequence_(RowSequence::CreateSequential(0, num_rows)),
+      null_flags_(BooleanChunk::Create(num_rows)) {
+  }
+
+  void Visit(const dhcore::column::CharColumnSource &source) final {
+    SimpleCopyValues<CharChunk, arrow::UInt16Builder>(source);
+  }
+
+  void Visit(const dhcore::column::Int8ColumnSource &source) final {
+    SimpleCopyValues<Int8Chunk, arrow::Int8Builder>(source);
+  }
+
+  void Visit(const dhcore::column::Int16ColumnSource &source) final {
+    SimpleCopyValues<Int16Chunk, arrow::Int16Builder>(source);
+  }
+
+  void Visit(const dhcore::column::Int32ColumnSource &source) final {
+    SimpleCopyValues<Int32Chunk, arrow::Int32Builder>(source);
+  }
+
+  void Visit(const dhcore::column::Int64ColumnSource &source) final {
+    SimpleCopyValues<Int64Chunk, arrow::Int64Builder>(source);
+  }
+
+  void Visit(const dhcore::column::FloatColumnSource &source) final {
+    SimpleCopyValues<FloatChunk, arrow::FloatBuilder>(source);
+  }
+
+  void Visit(const dhcore::column::DoubleColumnSource &source) final {
+    SimpleCopyValues<DoubleChunk, arrow::DoubleBuilder>(source);
+  }
+
+  void Visit(const dhcore::column::BooleanColumnSource &source) final {
+    SimpleCopyValues<BooleanChunk, arrow::BooleanBuilder>(source);
+  }
+
+  void Visit(const dhcore::column::StringColumnSource &source) final {
+    SimpleCopyValues<StringChunk, arrow::StringBuilder>(source);
+  }
+
+  void Visit(const dhcore::column::DateTimeColumnSource &source) final {
+    auto src_chunk = PopulateChunk<DateTimeChunk>(source);
+    auto dest_chunk = Int64Chunk::Create(num_rows_);
+    for (size_t i = 0; i != num_rows_; ++i) {
+      dest_chunk[i] = src_chunk[i].Nanos();
+    }
+    arrow::TimestampBuilder builder(arrow::timestamp(arrow::TimeUnit::NANO, "UTC"),
+        arrow::default_memory_pool());
+    PopulateAndFinishBuilder(dest_chunk, &builder);
+  }
+
+  void Visit(const dhcore::column::LocalDateColumnSource &source) final {
+    auto src_chunk = PopulateChunk<LocalDateChunk>(source);
+    auto dest_chunk = Int64Chunk::Create(num_rows_);
+    for (size_t i = 0; i != num_rows_; ++i) {
+      dest_chunk[i] = src_chunk[i].Millis();
+    }
+    arrow::Date64Builder builder;
+    PopulateAndFinishBuilder(dest_chunk, &builder);
+  }
+
+  void Visit(const dhcore::column::LocalTimeColumnSource &source) final {
```

Comment:
These three methods, Visit(DateTime), Visit(LocalDate) and Visit(LocalTime) have a lot of common structure, did you decide to avoid trying a template here for clarity?   Not a strong suggestion since sometimes a template may end up making the code harder to read and unnecessarily more complicated (specially since here we have 3 cases and not 20) but asking in terms of maybe thinking about it if you didn't.  Your call.

Reply:
Thank you for this. This class had gotten to be quite a mess so I just reorganized the whole thing. I think it's a lot more regular and consistent now. PTAL